src/optimizers/optimizer.cpp

```diff
@@ -29,186 +29,177 @@
+  // Item 3, p. 23 in "Effective C++", 3rd ed., by Scott Meyers
+  return const_cast<weights&>(static_cast<const optimizer&>(*this).get_weights());
+}
+
+const weights& optimizer::get_weights() const {
+  if (m_weights == nullptr) {
+    LBANN_ERROR("attempted to access the weights being optimized "
+                "before they are set");
   }
   return *m_weights;
 }
 
-const AbsDistMat& optimizer::get_gradient() {
+AbsDistMat& optimizer::get_gradient() {
 
-  // Check if gradient is initialized
-  if (!is_initialized()) {
-    LBANN_ERROR("attempted to access gradients before they are set up");
+  // Make sure gradient matrix has been setup
+  if (m_gradient == nullptr) {
+    LBANN_ERROR("attempted to access gradient before it is set up");
   }
 
-  // Perform allreduce on staging matrix if needed
-  if (m_gradient_allreduce_needed && !m_gradient_allreduce_started) {
-    start_gradient_staging_allreduce();
-  }
-  if (m_gradient_allreduce_started && !m_gradient_allreduce_finished) {
-    m_comm->wait(m_gradient_allreduce_req);
-    m_gradient_allreduce_finished = true;
+  // Make sure gradient values are ready
+  start_gradient_allreduce();
+  finish_gradient_allreduce();
+  if (m_gradient_status == optimizer_gradient_status::cleared) {
+    El::Zero(*m_gradient);
+    m_gradient_status = optimizer_gradient_status::ready;
   }
-  if (m_gradient_allreduce_needed) {
-    add_to_gradient(*m_gradient_staging);
+  if (m_gradient_status != optimizer_gradient_status::ready) {
+    std::ostringstream err;
+    err << "unexpected gradient status (expected "
+        << "\"" << to_string(optimizer_gradient_status::ready) << "\", "
+        << "but found \"" << to_string(m_gradient_status) << "\")";
+    LBANN_ERROR(err.str());
   }
-  m_gradient_allreduce_needed = false;
-  m_gradient_allreduce_started = false;
-  m_gradient_allreduce_finished = false;
 
+  // Return gradient
   return *m_gradient;
 
 }
 
-void optimizer::start_gradient_staging_allreduce() {
-  if (!m_gradient_allreduce_needed || m_gradient_allreduce_started) {
-    return;
-  }
-
-  m_gradient_allreduce_started = true;
-  m_comm->nb_allreduce(*m_gradient_staging,
-                       m_gradient_staging->RedundantComm(),
-                       m_gradient_allreduce_req,
-                       El::mpi::SUM);
-  m_gradient_allreduce_finished = false;
-}
-
-void optimizer::clear_gradient() {
-
-  // Clear matrices
-  El::Zero(*m_gradient);
-
-  // Reset gradient allreduce flags
-  m_gradient_allreduce_needed = false;
-  m_gradient_allreduce_started = false;
-  m_gradient_allreduce_finished = false;
-
-}
-
 void optimizer::add_to_gradient(const AbsDistMat& gradient,
-                                DataType scale) {
-  if (!is_initialized()) {
-    LBANN_ERROR("attempted to access gradients before they are set up");
+                                DataType scale,
+                                bool allreduce_needed) {
```

Comment:
I think your changes to this method result in us no longer doing non-blocking allreduces for gradient accumulation. All the layers with parameters call `add_to_gradient` with `allreduce_needed` as `true`, and so trigger an immediate blocking allreduce.

Reply:
`allreduce_needed` doesn't trigger the allreduce immediately, but just let's the optimizer know that one is needed before returning the gradient in `get_gradient()`. Most of the time, the non-blocking allreduce still happens in the same place (immediately after gradient contributions have been received from the appropriate layers during backprop).